allow const x = require('y') in local scope

kellyselden opened the pull request:
This would allow us to use this in ember-cli as mentioned [here](https://github.com/ember-cli/ember-cli/pull/7454#issuecomment-346244766).

I'm open to making this an optional thing, or you could deny if this goes against this package's purpose.

Commit:
allow const x = require('y') in local scope

rwjblue commented:
The idea here is that imports (requires in node-land) are _actual_ constants and cannot change.  Generally speaking these are done in module scope (and are therefore allowed as `const`), but in certain circumstances an import needs to be done lazily. In ember-cli we do this quite a bit to speed up command line performance where requiring everything in module scope would mean that we have to load the _entire possible program_ before continuing with executing the command which may only need a very small amount of the actual code (imagine loading all commands, tasks, build tooling, etc just for `ember version`).

This same rule applies to code that runs in a SPA, there are absolutely times where you need to dynamically import something. Those imports (using either ember-cli's global `require`, or the new [dynamic import api](https://github.com/tc39/proposal-dynamic-import) which is in TC39's stage 3) should be considered constant regardless of scoping rules...

rwjblue commented:
@kellyselden - Can you update this to account for [dynamic import](https://github.com/tc39/proposal-dynamic-import) as well?

kellyselden commented:
@rwjblue That seems a little out of scope of this IMO. I'm not familiar with the syntax. Doesn't it just return a promise? or is the desired syntax `const x = await import('y');`?

I think we should explore in a followup PR, (partly because it's a little fuzzy what the fix would look like).

cowboyd commented:
@kellyselden This seems right and proper.

Question: Are there cases where you do this _and_ are operating in an ES module environment?

If not, then perhaps it should only perform this check inside of a commonjs environment. Is there even a way to detect that?

cowboyd commented:
Also, if we're allowing `const` here as good style, should it be an error to say `let x = require('y');`? Seems odd for a plugin purporting to prefer `let` actually encourage the usage of `const` in certain cases. Nevertheless....

kellyselden commented:
I think it is still valid to do a lazy `require` in an ES module environment, since dynamic imports are such a recent addition.

As for your other questions, that makes me think this should be maybe an opt-in option, since it conflicts with the name of this plugin?

cowboyd commented:
In order to pick this back up, I like the idea of making it an option to start out with, and if it becomes too onerous, we can make it the default.

kellyselden commented:
I finally picked this back up. It is now opt-in via option `ignoreRequire`.

kellyselden commented:
Maybe I should change it to `ignoreModules` to cover both require and dynamic import?

cowboyd commented:
@kellyselden Yes, I like that. The only thing I think missing before merge is documentation on why you would want to allow requires or dynamic imports at any level.

Also, made you a collaborator so I don't block a merge when you're comfortable that it's ready.

kellyselden commented:
Updated the option name to `importModules` and added explanation. I left out dynamic imports for now because I think I would have to bump the min node version to start linting `const module = await import(moduleSpecifier);`. What do you think?

kellyselden commented:
Also you may want to add me to npm too.

Commit:
support dynamic imports

kellyselden commented:
I added a commit to support dynamic imports, but it may be too heavy. Let me know what you think.

MartinKolarik commented:
Any chance this could be merged?

cowboyd commented:
@kellyselden I'm ok with this if you want to go ahead and merge 👍

cowboyd commented:
Closing this as this repository has moved.. Feel free to open on https://github.com/thefrontside/javascript
